File: crawdad_tutorials/project2/project2_h2o.cpp
#include <cstdio>
#include <cmath>
#include "../molecule.hpp"

int main()
{
  /** Step 1: Read the Coordinate Data
   *
   */
  FILE *xyzfile;
  xyzfile = fopen("h2o_geom.txt", "r");
  int natom;
  fscanf(xyzfile, "%d", &natom);
  Molecule mol(natom, 0);
  for (int i = 0; i < mol.natom; i++)
    fscanf(xyzfile, "%lf %lf %lf %lf", &mol.zvals[i], &mol.geom[i][0], &mol.geom[i][1], &mol.geom[i][2]);
  fclose(xyzfile);

  /** Step 2: Read the Cartesian Hessian Data
   *
   */
  FILE *hessfile;
  hessfile = fopen("h2o_hessian.txt", "r");
  int hessatom;
  fscanf(hessfile, "%d", &hessatom);
  if (fabs(natom-hessatom) > 0) {
    printf("The number of atoms doesn't match.\n");
    return -1
  }
  double **H = new double* [3*natom];
  for (int i = 0; i < 3*natom; i++)
    new double[3*natom];
  for (int i = 0; i < 3*natom; i++)
    for (int j = 0; j < natom; j++)
      fscanf(hessfile, "%lf %lf %lf", &H[i][3*j], &H[i][3*j+1], &H[i][3*j+2]);
  fclose(hessfile);

  /** Step 3: Mass-Weight the Hessian Matrix
   * Divide each element of the Hessian matri by the product of square roots of the masses of the atoms associated with the given coordinates:
   *  \vect{F}_{M}^{ij} = \frac{F_{ij}}{\sqrt{m_{i}m_{j}}}
   */
  

  /** Step 4: Diagonalize the Mass-Weighted Hessian Matrix
   * Compute the eigenvalues of the mass-weighted Hessian:
   *  \vect{F}^{M}\vect{L} = \vect{L}\vect{\Lambda}
   */
  

  /** Step 5: Compute the Harmonic Vibrational Frequencies
   * The vibrational frequencies are proportional to the square root of the eigenvalues of the mass-weighted Hessian:
   *  \omega_{i} = \textrm{constant} \times \sqrt{\lambda_{i}}
   */

  return 0;
}

File: project2/project2_h2o.py
from __future__ import division

import math
import numpy as np
import scipy.linalg as spl

from crawdad_tutorials.utils import *

"""
Step 0: Read the Coordinate Data (piratechem)
"""
import piratechem as pc
h2o = pc.molecule.Molecule("water")
pc.read.xyz("h2o_geom.xyz", h2o)
natom = h2o.num_atoms()

"""
Step 1: Read the Coordinate Data
"""

"""
Step 2: Read the Cartesian Hessian Data
"""
hessfile = open("h2o_hessian.txt", "r")
hessatom = int(hessfile.readline())
H = np.zeros((3*natom, 3*natom))
for i in range(0,3*natom):
    for j in range(0,natom):
        line = hessfile.readline().split()
        H[i][3*j] = float(line[0])
        H[i][3*j+1] = float(line[1])
        H[i][3*j+2] = float(line[2])

print("Hessian:")
print_mat(H)
print("")

"""
Step 3: Mass-Weight the Hessian Matrix
"""
Hmw = np.zeros((3*natom, 3*natom))
for i in range(0,natom):
    for j in range(0,natom):
        mi = h2o.atoms[i].m
        mj = h2o.atoms[j].m
        mimj = math.sqrt(mi*mj)
        Hmw[i*natom+0][j*natom+0] = H[i*natom+0][j*natom+0]/mimj
        Hmw[i*natom+0][j*natom+1] = H[i*natom+0][j*natom+1]/mimj
        Hmw[i*natom+0][j*natom+2] = H[i*natom+0][j*natom+2]/mimj
        Hmw[i*natom+1][j*natom+0] = H[i*natom+1][j*natom+0]/mimj
        Hmw[i*natom+1][j*natom+1] = H[i*natom+1][j*natom+1]/mimj
        Hmw[i*natom+1][j*natom+2] = H[i*natom+1][j*natom+2]/mimj
        Hmw[i*natom+2][j*natom+0] = H[i*natom+2][j*natom+0]/mimj
        Hmw[i*natom+2][j*natom+1] = H[i*natom+2][j*natom+1]/mimj
        Hmw[i*natom+2][j*natom+2] = H[i*natom+2][j*natom+2]/mimj

print("Mass-weighted Hessian:")
print_mat(Hmw)
print("")

"""
Step 4: Diagonalize the Mass-Weighted Hessian Matrix
"""
omega = spl.eigvalsh(Hmw)

print("Mass-weighted Hessian eigenvalues:")
for i in omega:
    print("{0:12.10f}".format(i))
print("")

"""
Step 5: Compute the Harmonic Vibrational Frequencies
"""
freq = np.sqrt(omega) * vib_constant

print("Harmonic vibrational frequencies [cm]^-1:")
for i in freq:
    print("{0:10.4f}".format(i))
print("")
